src/prefect/server/database/dependencies.py

```diff
@@ -147,34 +184,167 @@ def db_injector(
 
     Returns:
-        A wrapped function with the PrefectDBInterface instance injected as the
-        first argument, preserving the original function's other parameters and
-        return type.
+        A wrapped descriptor object which injects the PrefectDBInterface instance
+        as the first argument to the function or method. This handles method
+        binding transparently.
+
     """
+    return DBInjector(func)
+
+
+# Descriptor object responsible for injecting the PrefectDBInterface instance.
+# It has no docstring so that
+class DBInjector(Generic[T, P, R]):
+    __slots__ = ("_func", "__name__")
+
+    _func: Union[_DBMethod[T, P, R], _DBFunction[P, R]]
+    __name__: str
+
+    if TYPE_CHECKING:
+
+        @overload
+        def __new__(cls, func: _DBMethod[T, P, R]) -> "DBInjector[T, P, R]":
+            ...
+
+        @overload
+        def __new__(cls, func: _DBFunction[P, R]) -> "DBInjector[None, P, R]":
+            ...
 
-    @wraps(func)
-    def sync_wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+        def __new__(
+            cls, func: Union[_DBMethod[T, P, R], _DBFunction[P, R]]
+        ) -> Union["DBInjector[T, P, R]", "DBInjector[None, P, R]"]:
+            ...
+
+    def __init__(self, func: Union[_DBMethod[T, P, R], _DBFunction[P, R]]) -> None:
+        super().__setattr__("_func", func)
+
+    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R:
         db = provide_database_interface()
+        func = cast(_DBFunction[P, R], self._func)
         return func(db, *args, **kwargs)
 
-    @wraps(func)
-    async def async_wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+    def __set_name__(self, owner: type[T], name: str) -> None:
+        super().__setattr__("__name__", name)
+
+        if issubclass(owner, BaseModel) and name in owner.__private_attributes__:
+            # Pydantic has marked this descriptor as a private attribute, this
+            # happens to all not explicitly registered descriptors.
+            # - Remove our decorated function from the private attributes
+            # - Update the class namespace to add this descriptor back in.
+            # - Update PrefectBaseModel 'ignored_types' to include this class so
+            #   all future pydantic models can avoid this issue.
+            from prefect._internal.schemas.bases import PrefectBaseModel
+
+            owner.__private_attributes__.pop(name, None)
+            setattr(owner, name, self)
+            ignored = PrefectBaseModel.model_config.get("ignored_types", ())
+            if DBInjector not in ignored:
+                PrefectBaseModel.model_config["ignored_types"] = (*ignored, DBInjector)
+
+            # TODO: what would it take for PrefectBaseModel to list this type as
+            # ignored, by default? This decorator is confined to the prefect.server
+            # namespace and PrefectBaseModel needs to work in a client-only context,
+            # so we rather not import across that boundary.
```

Comment:
There are two `PrefectBaseModel` implementations. One from `prefect._internal.schemas.bases` and one from ``prefect.server.utilities.schemas.bases`. The latter is the one that should be used server-side and I think that we could ignore this type by default there without crossing the client/server boundary.

Reply:
That is very interesting and would indeed make this much easier to resolve. I have a circular import to resolve though. Tomorrow.

Reply:
Resolved by creating a base descriptor class in the same module. Any subclass of that base is automatically included in the ignored types, and can trivially be imported without dependency issues.